### DISCUSSION TITLE
uv_tcp_close_reset() not calling uv_close() when setsockopt(3C) returns EINVAL

### DISCUSSION THREAD

**Discussion by alls0rts** [libuv/libuv#3658] [General]
Hi,

Seeing error `UV_EINVAL` returned from [uv_tcp_close_reset()](https://github.com/libuv/libuv/blob/v1.x/src/unix/tcp.c#L314) when a connection has been shut down.   The error is not from the initial test (`handle->flags & UV_HANDLE_SHUTTING`) but a return code from `setsockopt()`.

Looking at Solaris' man page to [setsockopt(3C)](https://docs.oracle.com/cd/E88353_01/html/E37843/setsockopt-3c.html) and the Open Group Specification ([Issue 7, 2018 edition](https://pubs.opengroup.org/onlinepubs/9699919799/toc.htm)) an `errno` of `EINVAL` can indicate that the socket has been shutdown already.

The error was observed while testing BIND 9.18.4 on Solaris 11.4 which asserts that this is an unexpected error.  This is logged with ISC as issue [#3283](https://gitlab.isc.org/isc-projects/bind9/-/issues/3283).

Should the correct course of uv_tcp_close_reset() be to go ahead and call uv_close() upon such an error, to continue to mark the socket closed and handle any callback?

Currently testing with a simple patch for `__sun`, but as a standard error it should perhaps be for all.

```
diff -r 927cb36365b3 src/unix/tcp.c
--- a/src/unix/tcp.c    Wed Jun 22 03:38:33 2022 -0700
+++ b/src/unix/tcp.c    Mon Jun 27 06:14:51 2022 -0700
@@ -321,7 +321,22 @@
 
   fd = uv__stream_fd(handle);
   if (0 != setsockopt(fd, SOL_SOCKET, SO_LINGER, &l, sizeof(l)))
+#if defined(__sun)
+  {
+    /*
+     * Open Group Specifications Issue 7, 2018 edition states that
+     * EINVAL may mean the socket has been shut down already.
+     * Call through to uv_close.
+     */
+     if (errno == EINVAL) {
+       errno=0;
+     } else {
+       return UV__ERR(errno);
+     }
+  }
+#else
     return UV__ERR(errno);
+#endif
 
   uv_close((uv_handle_t*) handle, close_cb);
   return 0;
```

After discussion I can followup with Issue and Pull request as needed.

Thanks for thoughts.

Stace

**Comment by vtjnash** [comment]
Sounds good to me

**Reply by alls0rts** [reply]
On the **BIND** discussion, it has been suggested that the same issue is seen on **macOS**.
I shall rework with generic changes, no C preprocess guards.

**Comment by alls0rts** [comment]
This has been addressed by pull https://github.com/libuv/libuv/pull/3662
